### PR TITLE
8328316: Finisher cannot emit if stream is sequential and integrator returned false

### PR DESCRIPTION
**Pull request by viktorklang-ora**
Adds differentiation between direct and transitive short circuiting which could prevent pushing downstream in the finisher for built-ins that were not `collect()`.

Creating this as a draft PR for now, just need to run some benchmarks to validate no significant regressions first.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8328316](https://bugs.openjdk.org/browse/JDK-8328316): Finisher cannot emit if stream is sequential and integrator returned false (**Bug** - P4)

### Reviewers
 * [Paul Sandoz](https://openjdk.org/census#psandoz) (@PaulSandoz - **Reviewer**) ⚠️ Review applies to [8974d9f9](https://git.openjdk.org/jdk/pull/18351/files/8974d9f9da703f953cc6fee6c989a3b42f484e20)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18351/head:pull/18351` \
`$ git checkout pull/18351`

Update a local copy of the PR: \
`$ git checkout pull/18351` \
`$ git pull https://git.openjdk.org/jdk.git pull/18351/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18351`

View PR using the GUI difftool: \
`$ git pr show -t 18351`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18351.diff">https://git.openjdk.org/jdk/pull/18351.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18351#issuecomment-2010559086)

### CODE REVIEW

**Comment by viktorklang-ora** [comment]
Just a note to self, consider if this should get backported to 22 or not.

**Comment by viktorklang-ora** [comment]
@PaulSandoz Would you mind reviewing this PR, Paul?

File: test/jdk/java/util/stream/GathererShortCircuitTest.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+public class GathererShortCircuitTest {
+    @Test
+    public void mustBeAbleToPushFromFinisher() {
+        Integer expected = 8328316;
+        List<Integer> source = List.of(1,2,3,4,5);
+
+        Gatherer<Integer, ?, Integer> pushOneInFinisher =
+                Gatherer.of(
+                    (unused, element, downstream) -> false,
+                    (unused, downstream) -> downstream.push(expected)
```
Comment (PaulSandoz) [inline test/jdk/java/util/stream/GathererShortCircuitTest.java:48]:
```suggestion
                    (_, element, downstream) -> false,
                    (_ downstream) -> downstream.push(expected)
```

**Review by PaulSandoz — APPROVED** [review]
Looks good, just a minor suggestion.

**Review by PaulSandoz — APPROVED** [review]

File: src/java.base/share/classes/java/util/stream/GathererOp.java [hunk 2]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
```
Comment (PaulSandoz) [inline src/java.base/share/classes/java/util/stream/GathererOp.java:2]:
```suggestion
 * Copyright (c) 2023, 2024, Oracle and/or its affiliates. All rights reserved.
```

**Pull request closed without merging**